apps/server/src/modules/user-migration/service/user-migration.service.ts

```diff
@@ -113,9 +129,41 @@ export class UserMigrationService {
 		return combinedUri;
 	}
 
-	private getMigrationRedirectUri(systemId: string): string {
-		const combinedUri = new URL(this.hostUrl);
-		combinedUri.pathname = `/api/v3/sso/oauth/${systemId}/migration`;
+	getMigrationRedirectUri(systemId: string): string {
+		const combinedUri = new URL(this.apiUrl);
+		combinedUri.pathname = `api/v3/sso/oauth/${systemId}/migration`;
 		return combinedUri.toString();
 	}
+
+	async migrateUser(currentUserId: string, externalUserId: string, targetSystemId: string): Promise<UserMigrationDto> {
+		const userDO: UserDO = await this.userService.findById(currentUserId);
+		const account: AccountDto = await this.accountService.findByUserIdOrFail(currentUserId);
+		const userDOCopy: UserDO = { ...userDO };
+		const accountCopy: AccountDto = { ...account };
+
+		try {
+			userDO.previousExternalId = userDO.externalId;
+			userDO.externalId = externalUserId;
+			userDO.lastLoginSystemChange = new Date();
+			await this.userService.save(userDO);
+			account.systemId = targetSystemId;
+			await this.accountService.save(account);
+
+			const userMigrationDto: UserMigrationDto = new UserMigrationDto({
+				redirect: `${this.hostUrl}/migration/succeed`,
+			});
+			return userMigrationDto;
+		} catch (e) {
+			await this.userService.save(userDOCopy);
+			await this.accountService.save(accountCopy);
+
+			this.logger.log(`This error occurred during migration of User: ${currentUserId} `);
```

Comment:
these messages absolutely need to be combined into one, otherwise their connection will get lost in the monitoring. (we dont have traceids yet, and even if we had them, its effort to gather messages togather based on a traceid).
the logger supports passing an object, you could do something like this
```
this.logger.log({
    message: 'error occured during migration of user, rolling back...',
    affectedUserId: currentUserId,
    error: e,
});